common/modules/gcp-project/main.tf

```diff
@@ -0,0 +1,105 @@
+# this Terraform module creates the same resources as project_init script and
+# set the DNS zones structure.
+
+terraform {
+  backend "gcs" {}
+}
+
+variable "project_name" {}
+variable "project_owner" {}
+variable "billing_id" {}
+variable "organization_id" {}
+
+locals  {
+  dnsname = "${replace(
+              replace(
+                google_project.project.name,
+                "/([a-z]+)-([a-z]+)-([a-z]+)-?([a-z]+)?/",
+                "$4.$3.$2.$1.net."),
+              "/^\\./",
+              "")
+            }"
+}
+
+resource "random_string" "id" {
+  length      = 4
+  min_numeric = 4
+}
+
+resource "google_project" "project" {
+  name            = "gpii-gcp-${var.project_name}"
+  project_id      = "gpii-gcp-${var.project_name}-${random_string.id.result}"
+  billing_account = "${var.billing_id}"
+  org_id          = "${var.organization_id}"
+}
+
+resource "google_project_services" "project" {
+  project = "${google_project.project.project_id}"
+  services = [
+    "oslogin.googleapis.com",
+    "compute.googleapis.com",
+    "logging.googleapis.com",
+    "monitoring.googleapis.com",
+    "appengine.googleapis.com",
+    "dns.googleapis.com"
+  ]
+}
+
+resource "google_service_account" "project" {
+  account_id   = "projectowner"
+  display_name = "Project owner service account"
+  project      = "${google_project.project.project_id}"
+}
+
+resource "google_project_iam_binding" "project" {
+  project = "${google_project.project.project_id}"
+  role    = "roles/owner"
+  members = [
+    "user:${var.project_owner}",
+    "group:ops@raisingthefloor.org",
+    "serviceAccount:${google_service_account.project.email}",
+    "serviceAccount:projectowner@gpii-common-prd.iam.gserviceaccount.com",
+  ]
+}
+
+resource "google_dns_managed_zone" "project" {
+  project     = "${google_project.project.project_id}"
+  name        = "${google_project.project.project_id}-zone"
+  dns_name    = "${local.dnsname}"
+  description = "${google_project.project.project_id} DNS zone"
+  depends_on  = ["google_project_services.project",
+                 "google_project_iam_binding.project"]
+}
+
+# Set the NS records in the parent zone of the parent project if the
+# project_name has the pattern ${env}-${user}
+resource "google_dns_record_set" "ns" {
+  name         = "${local.dnsname}"
+  managed_zone = "gpii-gcp-${element(split("-", var.project_name), 0)}-zone"
```

Comment:
The managed zones should have unique names as well, if we ensure that there won't be project duplication we can stay with this line untouched, otherwise a duplicated project will fail if the DNS zone name is already taken.

Reply:
We discussed this with @mrtyler and DNS names should be stable, as we're trying to hide the unique IDs from the user. You're not expected to ever have a duplicate project with the same name.